core/iwasm/libraries/debug-engine/debug_engine.c

```diff
@@ -109,73 +121,90 @@ control_thread_routine(void *arg)
     os_cond_signal(&debug_inst->wait_cond);
     os_mutex_unlock(&debug_inst->wait_lock);
 
-    /* wait lldb client to connect */
     if (!wasm_gdbserver_listen(control_thread->server)) {
-        LOG_ERROR("Failed while connecting debugger\n");
+        LOG_ERROR("Failed while listening for debugger\n");
         wasm_runtime_free(control_thread->server);
         return NULL;
     }
 
+    /* outer infinite loop: try to connect with the debugger */
     while (true) {
-        os_mutex_lock(&control_thread->wait_lock);
-        if (!should_stop(control_thread)) {
-            /* send thread stop reply */
-            if (debug_inst->stopped_thread
-                && debug_inst->current_state == APP_RUNNING) {
-                uint32 status;
-                korp_tid tid;
-
-                status =
-                    (uint32)
-                        debug_inst->stopped_thread->current_status->signal_flag;
-                tid = debug_inst->stopped_thread->handle;
-
-                if (debug_inst->stopped_thread->current_status->running_status
-                    == STATUS_EXIT) {
-                    /* If the thread exits, report "W00" if it's the last thread
-                     * in the cluster, otherwise ignore this event */
-                    status = 0;
-
-                    /* By design, all the other threads should have been stopped
-                     * at this moment, so it is safe to access the
-                     * exec_env_list.len without lock */
-                    if (debug_inst->cluster->exec_env_list.len != 1) {
-                        debug_inst->stopped_thread = NULL;
-                        /* The exiting thread may wait for the signal */
-                        os_cond_signal(&debug_inst->wait_cond);
-                        os_mutex_unlock(&control_thread->wait_lock);
-                        continue;
+        /* wait lldb client to connect */
+        if (!wasm_gdbserver_accept(control_thread->server)) {
+            LOG_ERROR("Failed while connecting debugger\n");
+            wasm_runtime_free(control_thread->server);
```

Comment:
`control_thread->server` will be freed in `wasm_debug_control_thread_destroy`, if you free it here, there may be double free issue

Reply:
Should I delete this line and another line above which free `control_thread->server` too? Or should I set `control_thread->server` to NULL after free?
above line:
```c
if (!wasm_gdbserver_listen(control_thread->server)) {
    LOG_ERROR("Failed while listening for debugger\n");
    wasm_runtime_free(control_thread->server);
    return NULL;
}
```

Reply:
Should delete these lines since the space for server will be freed during exiting